### ISSUE TITLE
Consider a license status helper or API

### ISSUE THREAD

**Issue by brockallen** [DuendeSoftware/IdentityServer#875]
https://github.com/DuendeSoftware/Support/issues/42

**Comment by brockallen** [comment]
And/or perhaps a flag to simulate a license tier for debug mode that informs you if you are violating a tier. Or a warning log entry?

**Comment by brockallen** [comment]
Or just a service in DI that reports the status, then people can built on top of that whatever they need/want.

**Comment by brockallen** [comment]
We still do want to improve this. For example, in dev, no license key -- how do they know what features they are using. We might want to emit a log entry in this scenario indicating that this feature is being used, so the license they would need would be "X".

**Comment by Pilchard123** [comment]
It would also be useful to know the maximum number of clients that a license allows, if the license in question is not unlimited (and maybe the number of clients used or the number still available).

**Comment by StuFrankish** [comment]
This is a bit of a bump, but I'm also interested in this - I'd like to make this part of our integrated Health Checks.
For example, given the capability a possible implementation could look like;
```c#
// Health Check Services
services.AddHealthChecks()
    .AddSqlServer(...)
    .AddIdentityServerLicenseCheck((license) => {
        license.LicenseKey = "...";
        license.FeatureFlags = ...;
    });
```
With a possible output of;
```
{
    "status": "Healthy",
    "totalDuration": "00:00:00.3611921",
    "entries": {
        "idsrv_license": {
            "data": {},
            "duration": "00:00:00.0226932",
            "status": "Healthy",
            "tags": []
        }
    }
}
```

**Comment by brockallen** [comment]
Any thoughts on this work: https://github.com/DuendeSoftware/IdentityServer/issues/1268

How would the ask in this issue change (if at all)?

**Comment by StuFrankish** [comment]
I think there's a fair amount of cross over between them - there's a lot of jumping around between issues in different repo's to mentally keep track of but #1268 which expands on #1266 covers the main point I at least would be interested in;

> Nice to have: A simple way to ask the IdentityServer Framework if the license is good.
>- A boolean to indicate if the license is valid
>- Optional: A message indicating the issue with the license
>- Optional: A accessible data structure like Duende.IdentityServer.Validation.License

After that, I'm happy to create an implementation of `IHealthCheck` and share for all interested parties.

**Comment by brockallen** [comment]
We'll see what we can do for our 7.0 release.

**Comment by Pilchard123** [comment]
I wonder if it would be helpful to have something to simulate having a particular license level for development. In #1305 you log when there's *no* license (and throw in some cases?), but it might be useful to be able to say - perhaps in development only - "pretend I have a license with these restrictions", so you don't get warnings for things that are going to be fixed by the license in production.

**Comment by brockallen** [comment]
If we made the License class public and put that in DI, would that give you what you need? If there's no license you'd get null. You could then build your own endpoints and/or health check logic based on whatever you needed/wanted based on what's important about the License to your scenarios.

Thoughts?

**Comment by brockallen** [comment]
So thinking more here... I guess I'm a bit hesitant to expose the full License class. I think we could make a new data structure that had the expiration, and license tier, and maybe the company name, and contact names/emails. Would there be anything else you want?

As for a flag for "is the license valid" that's hard since it depends on what people use. When something gets used that is not allowed in the license, then we emit errors or throw (now in v7.0) so I think that's the best we can do for those.

**Comment by StuFrankish** [comment]
> Would there be anything else you want?

That more than covers what I would ideally want 👍
> As for a flag for "is the license valid" that's hard since it depends on what people use. When something gets used that is not allowed in the license, then we emit errors or throw (now in v7.0) so I think that's the best we can do for those.

Just a thought, but.. could those features be configured during startup and checked at startup?
Certain license level features turned on or set, then throw at usage time that the feature isn’t enabled. Then further throw at startup that the license doesn’t support certain configured features?
Just spitballing here though don’t crucify me if the idea is stupid 🤣

**Comment by brockallen** [comment]
>  Just spitballing here though don’t crucify me if the idea is stupid 🤣

Nah, I think I see what you're asking. It's like you want to say I expect to use feature X, and Y and then the license validation runs and says yea/nay if the actual license allows that.

**Comment by Pilchard123** [comment]
> we added logging (for v7.0 -- so not released yet) when you're in dev and have no license that lets you know what license level you need.

Ah, that's exactly the sort of thing I was talking about.

> As for a flag for "is the license valid" that's hard since it depends on what people use.

> Then further throw at startup that the license doesn’t support certain configured features

Spitballing likewise, but how hard would it be to have a function exposed that checks if a license is valid for a particular use? So rather than having just "is this license valid", you could ask "I want to use DPoP. Does my license allow it?" or "I want to add another client. Does my license allow it?". Maybe "When does my license expire?" would be useful as well, so applications can log/display UI for when a license needs renewing.

Perhaps that could mean that whether to log or throw on startup is configurable. An application that knows it will only ever use some subset of features can have throw-on-startup enabled to make sure it never throws in cases where a user is trying to do something, but applications that might have sections enabled or disabled based on license level can start up just fine and only throw if they actually try to use the unlicensed features.

**Comment by brockallen** [comment]
> Spitballing likewise, but how hard would it be to have a function exposed that checks if a license is valid for a particular use? So rather than having just "is this license valid", you could ask "I want to use DPoP. Does my license allow it?" or "I want to add another client. Does my license allow it?" and enable/disable UI and behaviour based on the result. Maybe "When does my license expire?" would be useful as well, so applications can log/display UI for when a license needs renewing.

So this would require exposing all the data on the License class. I guess we can do that... I was just worried about any breaking changes that might happen in the future. For example, we renamed the redistribution flag at one point (from "ISV").

**Comment by brockallen** [comment]
Ok, so I think I have this handled in the PR now. I have made the license available in DI. We have a product specific license class with all the flags pertinent for the product. It's in DI, tho, so you can't do your own custom validation in ConfigureServices (since that's the nature of DI). But you can in Configure and/or in a health check.

**Comment by Pilchard123** [comment]
> So this would require exposing all the data on the License class.

I realise this is probably going to be *really irritating*, but I was mostly thinking aloud about possible use cases so I can't necessarily give strong feedback on a lot of it. The only bits that I'd be likely be using in anger (at least for now) are `Expiration` and `ClientLimit`. I'll have a play around with it later this evening and let you know what I think, though, and see if I can be at least a little useful.

**Comment by brockallen** [comment]
No worries. Given that you are interested in client limit, this means to me that others will want all the other flags on the license, thus the approach to expose the whole thing.

**Comment by Pilchard123** [comment]
I will get to giving you some feedback on this - I forgot I had to travel over the weekend and didn't have chance to give this proper time. Sorry :(
